### ISSUE TITLE
Code explaination

### ISSUE THREAD

**Issue by minhhuynh173** [tnmchau/MachineLearning#1]
import numpy
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import Flatten
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D
from keras.optimizers import Adam
from keras.utils import np_utils

_#  load dataset for training and testing from mnist, "bộ dữ liệu đưa vào là mảng 
3D (60000,28,28) với X_training và (10000,28,28) ứng với X_test
(X_train, y_train), (X_test, y_test) = mnist.load_data()
# Reshaping to format which CNN expects (batch, height, width, channels) nên ta reshaper lại thành 4D với batch là số lượng dataset, height, width, màu ảnh) ở đây channel là greyscale
X_train = X_train.reshape(X_train.shape[0], X_train.shape[1], X_train.shape[2], 1).astype('float32')
X_test = X_test.reshape(X_test.shape[0], X_test.shape[1], X_test.shape[2], 1).astype('float32')
# normalize inputs from 0-255 to 0-1 "với mỗi pixel trên ả có value từ 0-255 ta sẽ quy tỉ lệ thành 0-1 bằng cách chia 255"_

X_train/=255
X_test/=255
_# one hot encode (ở output cuối cùng cần phải phân lọai ra chữ số trong 10 số từ 0-9 nên sẽ là multi-class classification,  do output sẽ là one hot encode nên  sử liệu đầu ra np_utils.to_categorical(đựoc converts a class vector (integers) to binary class matrix. )_ 

number_of_classes = 10
y_train = np_utils.to_categorical(y_train, number_of_classes)
y_test = np_utils.to_categorical(y_test, number_of_classes)
# create model
_**https://faroit.github.io/keras-docs/1.2.2/layers/pooling/  (nguồn tham khảo)**_
_ //passing a list of layer instances to the constructor_
model = Sequential()   _//khai bao sequential is a linear stack of layers

_//apply a 5x5 convolution with 32 output filters on a 28x28 image
first hidden layer ( 2D có (32 filter,size(5x5), input_shape= 28x28 , activation='relu'_
model.add(Conv2D(32, (5, 5), input_shape=(X_train.shape[1], X_train.shape[2], 1), activation='relu')) 

-//Pooling layer () reduce 5x5 -> 3x3 covolution
model.add(MaxPooling2D(pool_size=(2, 2)))

_//apply a 3x3 convolution with 32 output filters on a 3x3 image, activation là ReLU_ 
model.add(Conv2D(32, (3, 3), activation='relu'))

_//Pooling layer () reduce 3x3 -> 2x2 covolution_
model.add(MaxPooling2D(pool_size=(2, 2)))

_// _
model.add(Dropout(0.2))

_//converts the 2D matrix data to a vector  _
model.add(Flatten())

_// layer is a fully connected layer with 128 neurons_
model.add(Dense(128, activation='relu'))

_// output layer with 10 neurons(number of output classes), uses softmax activation function _
model.add(Dense(number_of_classes, activation='softmax'))

# Compile model
model.compile(loss='categorical_crossentropy', optimizer=Adam(), metrics=['accuracy'])
# Fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=200)
# Save the model
model.save('models/mnistCNN.h5')
# Final evaluation of the model
metrics = model.evaluate(X_test, y_test, verbose=0)
print("Metrics(Test loss & Test Accuracy): ")
print(metrics)
